### ISSUE TITLE
.Net: Function calling is producing inconsistent results

### ISSUE THREAD

**Issue by gmantri** [microsoft/semantic-kernel#4441]
**Describe the bug**
We are building a RAG Bot where users can ask questions based on internal company documents. The questions could be related to technology, policies etc. For each kind of question (technology, policy), we have created different semantic functions (using YAML). We also have some native functions as well.

When a user asks a question, we first determine what kind of question is that (technology etc.) and that is working great. Now we want to find out the function we want to invoke (we are opting for manual function invocation), and for that we are using function calling feature of Open AI.

What we are seeing is that for the same question and question kind, we are getting inconsistent results. For example, if we ask a question "What is Azure OpenAI" (which is a technology kind of question), sometimes we get no matching function, at times we get `technology` function and then at times we get `policy` function back.

Here's the code we wrote:

```
private async Task<OpenAIFunctionToolCall> SelectTool(string userask, string type)
{
	var kernel = GetKernel();//Get kernel with all plugins and functions loaded
	var prompt = $"User ask is '{userask}' and the type is '{type}'. Select the best function to answer user question based on the type.";
	var chatCompletionService = new AzureOpenAIChatCompletionService("mymodelname", "https://xyz.openai.azure.com/", new DefaultAzureCredential());
	var result = await chatCompletionService.GetChatMessageContentAsync(new ChatHistory(prompt),
		new OpenAIPromptExecutionSettings()
		{
			ToolCallBehavior = ToolCallBehavior.EnableKernelFunctions
		}, kernel);
	var functionCall = ((OpenAIChatMessageContent)result).GetOpenAIFunctionToolCalls().FirstOrDefault();
	return functionCall;
}
```

**Expected behavior**
Given same question and type, we expect to get the same function back.

**Platform**
 - OS: Windows
 - IDE: Visual Studio
 - Language: C#
 - Source: Microsoft.SemanticKernel (version 1.0.1), Microsoft.SemanticKernel.PromptTemplates.Handlebars (version 1.0.1), Microsoft.SemanticKernel.Yaml (version 1.0.1), Azure.AI.OpenAI (version 1.0.0-beta.12).

**Comment by gmantri** [comment]
I think I know why this is happening. My guess is that it is happening because I am not explicitly setting the `temperature` to `0`. 

I created a simple console application that makes use of Math plugin from [`here`](https://learn.microsoft.com/en-us/semantic-kernel/agents/plugins/using-the-kernelfunction-decorator?tabs=Csharp#creating-the-remaining-math-functions) and asked a simple question (`What is 2 + 2`). Ran the code in loop for 1000 times. With no temperature setting, I saw weird results. Then I set the temperature to 0 and reran the code in loop for 1000 times. This time I got the correct results.

Can someone here confirm that it is indeed the cause for this weird behavior?

Here's my code:

```
using Azure.AI.OpenAI;
using Microsoft.SemanticKernel;
using Microsoft.Extensions.Logging;
using Plugins;
using Microsoft.SemanticKernel.Connectors.OpenAI;
using Microsoft.SemanticKernel.ChatCompletion;

const string AzureOpenAIEndpoint = "https://abc.openai.azure.com/";
const string AzureOpenAIKey = "my-super-secret-key";
const string AzureOpenAIModelId = "chatbot-gpt-35";


var openAIClient = new OpenAIClient(new Uri(AzureOpenAIEndpoint), new Azure.AzureKeyCredential(AzureOpenAIKey));

var prompt = "What is 2 + 2?";

for (var i=0; i<1000; i++)
{
	var function = await SelectTool(prompt);
	
	Console.WriteLine($"{(i+1).ToString("d4")}) {prompt}: {function ?? "Not Found"}");
}

async Task<string?> SelectTool(string prompt)
{
	try
	{
		var kernel = GetKernel();
		var chatCompletionService = new AzureOpenAIChatCompletionService(AzureOpenAIModelId, openAIClient!);
		var result = await chatCompletionService.GetChatMessageContentAsync(new ChatHistory(prompt),
			new OpenAIPromptExecutionSettings()
			{
				ToolCallBehavior = ToolCallBehavior.EnableKernelFunctions,
				Temperature = 0
			}, kernel);
		var functionCall = ((OpenAIChatMessageContent)result).GetOpenAIFunctionToolCalls().FirstOrDefault();

		return functionCall?.ToString();
	}
	catch (Exception ex)
	{
		return "Error";
	}
}

Kernel GetKernel()
{
	var kernelBuilder = Kernel.CreateBuilder()
		.AddAzureOpenAIChatCompletion(AzureOpenAIModelId, openAIClient);
	
	var kernel = kernelBuilder.Build();
	kernel.Plugins.AddFromType<MathPlugin>();
	return kernel;
}
```

**Comment by crickman** [comment]
Thank you for sharing this experience.  indeterminacy is certainly one challenging aspect of incorporating AI features into applications.

You are correct that setting Temperature = 0 will maximize determinacy, but it isn't a guarantee.

Another critical aspect that contributes to function selection is the description metadata.  Very similiar to prompt engineering, you may need to experiment with some different approaches to describing your function and parameters.

In short, yes...temperature = 0 is the proper course of action.  (There may be other natural language scenarios where more breadth in the variety of results in desired...but this consideration doesn't necessarily extend to functional responses.

**Comment by gmantri** [comment]
@crickman - Thank you for the note. I completely agree with you that along with setting temperature to 0 is needed and also the need for the proper description of Kernel functions. With these two things, I was able to get predictable results in the tests that I have done so far.
